### DISCUSSION TITLE
Closed Loop example in position mode and velocity mode

### DISCUSSION THREAD

**Discussion by EhsanRasouli1371** [Peter-van-Tol/LiteX-CNC#97] [Q&A]
How to create a closed loop system in Litex?

**Comment by EhsanRasouli1371** [comment]
In speed mode:
HAL:
```
net x-enable      pid.x.enable
net x-enable      5A-75B:V8.2.stepgen.00.enable 
net x-enable      joint.0.amp-enable-out            5A-75B:V8.2.gpio.out23.out
net x-enable      5A-75B:V8.2.gpio.inablestep.out
                  
setp    5A-75B:V8.2.stepgen.00.velocity-mode 1	
setp    5A-75B:V8.2.encoder.00.x4-mode 1
setp    5A-75B:V8.2.encoder.00.position-scale  [JOINT_0]ENCODER_SCALE

net x-pos-fb      5A-75B:V8.2.encoder.00.position  =>  pid.x.feedback      
net x-pos-fb     => joint.0.motor-pos-fb

setp   pid.x.Pgain     [JOINT_0]P
setp   pid.x.Igain     [JOINT_0]I
setp   pid.x.Dgain     [JOINT_0]D
setp   pid.x.bias      [JOINT_0]BIAS
setp   pid.x.FF0       [JOINT_0]FF0
setp   pid.x.FF1       [JOINT_0]FF1
setp   pid.x.FF2       [JOINT_0]FF2
setp   pid.x.deadband  [JOINT_0]DEADBAND
setp   pid.x.maxoutput [JOINT_0]MAX_OUTPUT
setp   pid.x.error-previous-target  true
setp   pid.x.maxerror   0.5            

net x-pos-cmd      joint.0.motor-pos-cmd  =>  pid.x.command
net x-output        pid.x.output          => 5A-75B:V8.2.stepgen.00.velocity-cmd 
 
# Step Gen signals/setup

setp 5A-75B:V8.2.stepgen.00.position-scale   [JOINT_0]STEP_SCALE
setp 5A-75B:V8.2.stepgen.00.steplen          [JOINT_0]STEPLEN
setp 5A-75B:V8.2.stepgen.00.stepspace        [JOINT_0]STEPSPACE
setp 5A-75B:V8.2.stepgen.00.dir-hold-time    [JOINT_0]DIRHOLD
setp 5A-75B:V8.2.stepgen.00.dir-setup-time   [JOINT_0]DIRSETUP
setp 5A-75B:V8.2.stepgen.00.max-velocity     [JOINT_0]MAX_VELOCITY
setp 5A-75B:V8.2.stepgen.00.max-acceleration [JOINT_0]STEPGEN_MAXACCEL


net x-index-enable           pid.x.index-enable
net x-index-enable           joint.0.index-enable  <=>  5A-75B:V8.2.encoder.00.index-enable  
    
# ---setup home / limit switch signals---

net x-home-sw           => joint.0.home-sw-in
net x-home-sw          <= 5A-75B:V8.2.gpio.in18.in
```
and INI:
```
[AXIS_X]
MAX_VELOCITY = 9.9
MAX_ACCELERATION = 20
MIN_LIMIT = -1
MAX_LIMIT = 500.0

[JOINT_0]
TYPE = ANGULAR
HOME = 0.0
MIN_LIMIT = -1
MAX_LIMIT = 500.0

FERROR = 1
MIN_FERROR = 1

MAX_VELOCITY = 9.9
MAX_ACCELERATION = 20
STEPGEN_MAXVEL = 9.9
STEPGEN_MAXACCEL = 30

SCALE = 10000

HOME_OFFSET = 0.000000
HOME_SEQUENCE = 1

P = 20
I = 0
D = 0
FF0 = 0
FF1 = 0
FF2 = 0
BIAS = 0
DEADBAND = 0.00
MAX_OUTPUT = 0
ENCODER_SCALE = 2500
DIRSETUP   = 1200
DIRHOLD    = 1200
STEPLEN    = 1000
STEPSPACE  = 1000
STEP_SCALE = 10000
HOME_SEARCH_VEL = .6
HOME_LATCH_VEL = 0.1
HOME_FINAL_VEL = .3
HOME_USE_INDEX = YES
#VOLATILE_HOME = 1
HOME_NO_ENCODER_RESET = YES
HOME_SEQUENCE = 1

HOME = 0
```

**Reply by EhsanRasouli1371** [reply]
in position mode:
-encoder feedback must be wired and used by LinuxCNC
-enable output must be wired to drive enable inputs
HAL:
```
net xpos_cmd   joint.0.motor-pos-cmd => 5A-75B:V8.2.stepgen.00.position-cmd
net xpos_fb    joint.0.motor-pos-fb  <= 5A-75B:V8.2.encoder.00.position
net xenable joint.0.amp-enable-out => 5A-75B:V8.2.stepgen.00.enable 5A-75B:V8.2.gpio.inablestep.out  5A-75B:V8.2.gpio.out23.out
```

**Comment by EhsanRasouli1371** [comment]
Is the HAL configuration correct in both modes?

**Reply by Peter-van-Tol** [reply]
The system you describe is not a true closed loop example. A real closed loop example would take an input from and linear encoder and using that as feedback for LinuxCNC. Would love to try that out with a simple setup if somebody would be willing to sponsor the materials...

In your example you use the PID component to translate between position as commanded by LinuxCNC and velocity as required by the FPGA (**note**: there is no trajectory planner in the FPGA! The FPGA relies on a steady stream of new velocities to form the position profile. this is very close to the architecture of LinuxCNC stepgen component.). I have no experience in tuning these components and also believe that it might be machine dependent.

The FPGA does have a position mode. In this case the feedback loop is closed within the component. One of the known limitations is humming at very low speeds or stand-still. The reason for this is that the packages arrive at random times, i.e. the control loop of LinuxCNC has some jitter around it. When a position is commanded and held, there might be a small overshoot at each step, which is compensated and leads to a new overshoot, etc....

In LinuxCNC stepgen this is compensated by just doing the remaining step when the difference is small enough. Maybe I've to add the same functionality in the FPGA to prevent this humming. But this might require sending positions instead of velocities instead and a large re-write of the stepgen...

**Comment by EhsanRasouli1371** [comment]
hello peter
Can you please read this thread?
https://forum.linuxcnc.org/24-hal-components/48222-hal-help-for-home-to-index-with-stepgen-firmware-update
Can you create pin stepgen.00.index-enable ?

**Reply by Peter-van-Tol** [reply]
I can create the pin, that's one line of code. But can you define the function of this pin?
- Is it INPUT (going to stepgen) or OUTPUT (receiving from stepgen);
- When INPUT
  - When should the pin turn HIGH and when LOW
- When OUTPUT:
  - What should happen when the pin is HIGH and what when the pin is LOW

**Reply by hmnijp** [reply]
board-name.stepgen.nn.index_enable (HAL_BIT input)

this pin starts a procedure similar to .encoder.<n>.index_enable for the corresponding stepgen number. (encoder.02 reset to zero stepgen.02 etc..)

When stepgen.00.index_enable is true, the counts and position of stepgen.00 are reset to zero on the next rising edge Phase-Z of encoder.00.  At the same time, index-enable is reset to zero to indicate that a rising edge has occurred.

This functionality is used for homing using only Phase-Z, without using the feedback loop and AB phase of the encoder.

**Reply by Peter-van-Tol** [reply]
For the firmware the described solution is feasible. I'm thinking of the following:
- one can set the `index_enable` on the stepgen
- when the `index`-pin of the stepgen is HIGH, the `counter` is reset to zero and the `index_enable` for that channel is reset to `0`
- the new value of the `counter` and `index_enable` will be send to LinuxCNC at the next cycle.

Challenge is that the absolute position of the machine jumps when the `counter` of the stepgen is reset. I don't know whether the motion planner of LinuxCNC can handle this jump. Do you know this?

When it can handle the jump, the absolute position is equal to the relative position (relative as in relative to the home position).

**Reply by Peter-van-Tol** [reply]
After reading the manual for the MESA, I can build this feature I think. The changes are minimal, so I think I can change it this evening.

@hmnijp : Would you be able to test this feature?

Another question: I also see the pins `latch_enable` and `latch_position` in the MESA stepgen HAL. Are these pins also required in the implementation of the stepgen?

**Reply by hmnijp** [reply]
Yes, I have a motor with encoder and Z phase, I think I can test it in the next few days. 

This feature has really proven to be in demand:
Many people who use ac servos (all have abz encoder) want to use HOME_USE_INDEX (high precision home procedure), but not though to use closed loop to linuxcnc with velocity control mode and pid control to reset encoder.pos-fb using index. For this purpose it was invented to use resetting the stepgen counter with index Z bypassing the encoder counter.

> Another question: I also see the pins latch_enable and latch_position in the MESA stepgen HAL. Are these pins also required in the implementation of the stepgen?

I haven't seen it used at this time.

**Reply by Peter-van-Tol** [reply]
As a first step I will just zero the counter on the rising Z-index. Will implement this on Wednseday.

**Reply by Peter-van-Tol** [reply]
Issue #111 created to resolve this.

**Reply by Peter-van-Tol** [reply]
@hmnijp : I was programming this in the FPGA (not yet committed due to some issues), but was questioning myself whether this is necessary? When the `stepgen` is in velocity-mode, this module does not use the information about position. The feedback comes from the encoder.

Would it be possible:
- ignore the position feedback from the `stepgen`
- to only use the position of the `encoder` as feedback for `motmod` (see the required change below, it is already implemented) and the `pid`;

Why is the position required from two sources?

**Reply by hmnijp** [reply]
> When the `stepgen` is in velocity-mode, this module does not use the information about position. The feedback comes from the encoder.

stepgen can work in velocity mode using not only encoder.pos-fb, but also stepgen.pos-fb - in this case it is also necessary to reset stepgen.pos-fb (raw counts) together with z-index.

(I personally use this configuration because it has less f-error. PID allows to correct the acceleration with FF2 - this is not realized in the stepgen module with position control, so according to my tests - speed control is preferable, even without encoder-fb.)
![photo_2024-04-14_14-35-35](https://github.com/user-attachments/assets/42ffe18a-1091-4240-ab8e-8543e4b8aa4d)

Another use case is to control a servo spindle using stepdir.

**Reply by Peter-van-Tol** [reply]
At this moment I've implemented the `index-enable` by storing the position where the index was triggered. Before the data is send back to the CPU this index position is subtracted from the actual position. I cannot reset the counter to zero, as this would lead to an extra step if the index pulse happens at certain moments. So the `stepgen` is internally always using its own reference.

The `index-enable` signal is sending using the most significant byte of the acceleration command. This field is rarely used in full extent, and it safe a lot of rewriting the data. To get the index-flag back to the CPU I will use a similar track with the speed-feedback.

The firmware is done and tested in simulation. Tomorrow I will continue with the driver.

**Reply by hmnijp** [reply]
> by storing the position where the index was triggered. Before the data is send back to the CPU this index position is subtracted from the actual position

ok, this should be enough to set the joint motor offset correctly.

> The index-enable signal is sending using the most significant byte of the acceleration command. This field is rarely used in full extent, and it safe a lot of rewriting the data. To get the index-flag back to the CPU I will use a similar track with the speed-feedback.

Since for this configuration we need to use Encoder Instance (only Z, without AB) - is it possible to use its similar fields for this?

**Reply by Peter-van-Tol** [reply]
> Since for this configuration we need to use Encoder Instance (only Z, without AB) - is it possible to use its similar fields for this?

The configuration of the stepgen will be:
``` json
    "modules": [
        {
            "module_type": "stepgen",
            "instances": [
                {
                    "pins" : {
                        "stepgen_type": "step_dir",
                        "step_pin": "j7:0",
                        "dir_pin": "j7:1",
                        "index_pin": "j7:2"
                    },
                    "soft_stop": true
                }
             ]
         }
    ],
```

When the `index_pin` is set, it is possible to reset the counter with the `index_enable` pin in the HAL. When you want to re-use an existing encoder for it (i.e., you want both to use the index-pulse and also position feedback from the same encoder), you have to wire the Z-index to two pins of the FPGA.

**Reply by hmnijp** [reply]
I get the logic. It looks good!

**Reply by Peter-van-Tol** [reply]
I think the version is ready to be tested. This branch now contains:
- configurable maximum frequency (see #91 ), I needed to change the configuration fields. The maximum pick-off is now 16, but this is only required for very slow steppers. The resolution of the step = 0.001 steps/second with a maximum step frequency of 1 MHz with a 40 MHz clock. This increases to 0.0006 steps/second for the defaulkt400 kHz with a 40 MHz clock.
- individual timings per stepgen instance (see #91)
- the following extra pins are exported when the `stepgen` supports indexing
  - pin `index-enable`: set HIGH to enable indexing. When this pin is HIGH, the counter will be reset to 0 on a rising edge of `index_pin`. This function will only capture the first rising edge. This pin must set to LOW to clear the `index-pulse` flag and reset the indexing logic.
  - pin `index-pulse`: is HIGH when indexing was enabled and a rising edge has been detected. 
-  the resolution of the acceleration setting in the FPGA is improved with a factor of 256.

When this test goes well, I think both this issue and #91 can be closed.

**Reply by hmnijp** [reply]
thank you for your work!
I'll do some testing and come back with the results...

**Reply by Peter-van-Tol** [reply]
@hmnijp : did you have had any chance to test the changes?

**Comment by hmnijp** [comment]
And I remembered that I found another small bug concerning HOME_USE_INDEX that I forgot to write about:

when index is found, linuxcnc resets 

> joint.pos-cmd = 0 
> encoder.pos-fb = 0. 

At the end of the homing, position and feedback are zero, as they should be, there is no error.

But if HOME_OFFSET =/= 0 is used, then at the end of the procedure 

> joint.pos-cmd = 0 + home_offset
> encoder.pos-fb = 0

It's the difference between pos and feedback that causes the **follow error** after homing.

I think when using index-enable we should set 
`encoder.position_offset = home_offset`
e.g. make it constant in hal
`setp encoder.00.position_offset [JOINT_0]HOME_OFFSET `
then after resetting the encoder counter to 0 there will be no error because encoder.pos-fb = counter+position_offset.

This pin (encoder.00.position_offset) is mentioned in the documentation, but it doesn't seem to exist in the current version of the encoder component...


> and maybee after homing with index
> joint.pos-cmd = 0 + offset + **position-latched** ?
> I need to look for information about it...

**Reply by Peter-van-Tol** [reply]
You're correct that the `position_offset` is not exported to HAL. Will fix it with the feature request as well.

**Reply by Peter-van-Tol** [reply]
See #111

The `position-offset` is implemented in the HAL. The position is calculated as:
```
position = position-offset + counts / position-scale
```

Both `position-scale` and `position-offset` are parameters in HAL, not pins. If this is a problem, I can move them to the pins section.
